File: offload/test/offloading/fortran/error_directive.f90
! Test the `error` directive with `at(execution)` inside a target region.
!
! REQUIRES: flang, libc

! flang does not accept -foffload-lto.
! UNSUPPORTED: nvptx64-nvidia-cuda-LTO

! RUN: %libomptarget-compile-fortran-generic -fopenmp-version=51 && \
! RUN:   %libomptarget-run-generic 2>&1 | %fcheck-generic --check-prefix=NOLOC
! RUN: %libomptarget-compile-fortran-generic -fopenmp-version=51 -g && \
! RUN:   %libomptarget-run-generic 2>&1 | %fcheck-generic --check-prefix=LOC

program error_directive
  implicit none

  !$omp target
  !$omp error at(execution) severity(warning) message("warning message")
  !$omp end target

  ! No MESSAGE clause, so the runtime receives a null message pointer.
  !$omp target
  !$omp error at(execution) severity(warning)
  !$omp end target
end program error_directive

! Device output is flushed after host output, so host prints are not checked.
! Without -g the ident holds no location and is reported as "unknown:0:0", the
! same as the host runtime.

! NOLOC: OMP: unknown:0:0: Encountered user-directed warning: warning message.
! NOLOC: OMP: unknown:0:0: Encountered user-directed warning.

! LOC: {{.*}}error_directive.f90:{{[0-9]+}}:{{[0-9]+}}: Encountered user-directed warning: warning message.
! LOC: {{.*}}error_directive.f90:{{[0-9]+}}:{{[0-9]+}}: Encountered user-directed warning.

File: offload/test/offloading/error_directive.c
// Test the `error` directive with `at(execution)` inside a target region.
//
// REQUIRES: libc

// RUN: %libomptarget-compile-generic -fopenmp-version=51 && \
// RUN:   %libomptarget-run-generic 2>&1 | %fcheck-generic
// RUN: %libomptarget-compile-generic -fopenmp-version=51 -DFATAL && \
// RUN:   %libomptarget-run-fail-generic

#include <stdio.h>

int main(void) {
#ifdef FATAL
#pragma omp target
  {
#pragma omp error at(execution) severity(fatal) message("fatal message")
  }
  printf("unreachable\n");
#else
#pragma omp target
  {
#pragma omp error at(execution) severity(warning) message("warning message")
  }

  // No MESSAGE clause, so the runtime receives a null message pointer.
#pragma omp target
  {
#pragma omp error at(execution) severity(warning)
  }
#endif
  return 0;
}

// Device output is flushed after host output, so host prints are not checked.
// The fatal case checks only the exit status: its message is lost when the trap
// aborts before the buffered stdout is flushed.
//
// clang fills in the ident location from the AST, so it is present without -g

// CHECK: {{.*}}error_directive.c:{{[0-9]+}}:{{[0-9]+}}: Encountered user-directed warning: warning message.
// CHECK: {{.*}}error_directive.c:{{[0-9]+}}:{{[0-9]+}}: Encountered user-directed warning.
